File: recursion/powx-n.py
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if(n==0):
            return 1.00
        
        if(n < 0):
            x = 1/x
            n = -n
        
        half = self.myPow(x,n//2)
        if(n%2==0):
            return half*half
        return half*half*x

File: math/powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        long long N = n;
        if(N==0){
            return 1;
        }
        else if(N < 0){
            x = 1.0/x;
            N = -N;
        }
        double half = myPow(x,N/2);
        if(N%2==0){
            return half*half;
        }
        return x*half*half;
    }
};